codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/Paginators.java

```diff
@@ -0,0 +1,254 @@
+                .map(operationShape -> paginatedIndex.getPaginationInfo(serviceShape, operationShape))
+                .filter(Optional::isPresent)
+                .map(paginationInfo -> paginationInfo.get())
+                .collect(Collectors.toList());
+
+        if (paginationInfos.size() == 0) {
+            return;
+        }
+
+        writerFactory.accept("paginators.go", settings.getModuleName(), writer -> {
+            paginationInfos.forEach((paginationInfo) -> {
+                generateOperationPaginator(model, symbolProvider, writer, paginationInfo);
+            });
+        });
+    }
+
+    private void generateOperationPaginator(
+            Model model,
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            PaginationInfo paginationInfo
+    ) {
+        Symbol operationSymbol = symbolProvider.toSymbol(paginationInfo.getOperation());
+        Symbol inputSymbol = symbolProvider.toSymbol(paginationInfo.getInput());
+        Symbol outputSymbol = symbolProvider.toSymbol(paginationInfo.getOutput());
+        Optional<MemberShape> pageSizeMember = paginationInfo.getPageSizeMember();
+
+        Symbol interfaceSymbol = SymbolUtils.createValueSymbolBuilder(String.format("%sAPIClient",
+                operationSymbol.getName())).build();
+        Symbol paginatorSymbol = SymbolUtils.createPointableSymbolBuilder(String.format("%sPaginator",
+                operationSymbol.getName())).build();
+        Symbol optionsSymbol = SymbolUtils.createPointableSymbolBuilder(String.format("%sOptions",
+                paginatorSymbol.getName())).build();
+
+        writeClientOperationInterface(writer, operationSymbol, inputSymbol, outputSymbol, interfaceSymbol);
+        writePaginatorOptions(model, symbolProvider, writer, operationSymbol, pageSizeMember, optionsSymbol);
+        writePaginator(symbolProvider, writer, paginationInfo, operationSymbol, inputSymbol, outputSymbol,
+                pageSizeMember, interfaceSymbol, paginatorSymbol, optionsSymbol);
+    }
+
+    private void writePaginator(
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            PaginationInfo paginationInfo,
+            Symbol operationSymbol,
+            Symbol inputSymbol,
+            Symbol outputSymbol,
+            Optional<MemberShape> pageSizeMember,
+            Symbol interfaceSymbol,
+            Symbol paginatorSymbol,
+            Symbol optionsSymbol
+    ) {
+        String inputMember = symbolProvider.toMemberName(paginationInfo.getInputTokenMember());
+
+        writer.writeDocs(String.format("%s is a paginator for %s", paginatorSymbol, operationSymbol));
+        writer.openBlock("type $T struct {", "}", paginatorSymbol, () -> {
+            writer.write("options $T", optionsSymbol);
+            writer.write("client $T", interfaceSymbol);
+            writer.write("params $P", inputSymbol);
+            writer.write("nextToken *string");
```

Comment:
type of `nextToken` probably should be $P of symbol of next token member. Especially to support things like DDB's next token.